src/audio/eq_iir/eq_iir_ipc4.c

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2017-2022 Intel Corporation. All rights reserved.
+//
+// Author: Seppo Ingalsuo <seppo.ingalsuo@linux.intel.com>
+//         Liam Girdwood <liam.r.girdwood@linux.intel.com>
+//         Keyon Jie <yang.jie@linux.intel.com>
+
+#include "eq_iir.h"
+#include <sof/audio/component.h>
+#include <sof/audio/module_adapter/module/generic.h>
+#include <sof/audio/data_blob.h>
+#include <sof/audio/buffer.h>
+#include <sof/audio/format.h>
+#include <sof/audio/pipeline.h>
+#include <sof/audio/ipc-config.h>
+#include <sof/common.h>
+#include <rtos/panic.h>
+#include <sof/ipc/msg.h>
+#include <rtos/alloc.h>
+#include <rtos/init.h>
+#include <sof/lib/memory.h>
+#include <sof/lib/uuid.h>
+#include <sof/list.h>
+#include <sof/math/iir_df1.h>
+#include <sof/platform.h>
+#include <rtos/string.h>
+#include <sof/ut.h>
+#include <sof/trace/trace.h>
+#include <ipc/control.h>
+#include <ipc/stream.h>
+#include <ipc/topology.h>
+#include <user/eq.h>
+#include <user/trace.h>
+#include <errno.h>
+#include <stddef.h>
+#include <stdint.h>
+
+LOG_MODULE_DECLARE(eq_iir, CONFIG_SOF_LOG_LEVEL);
+
+static eq_iir_func eq_iir_find_func(struct processing_module *mod)
```

Comment:
Also add comment that ipc4 functions finding is different because of different formats handling capability. I think nothing technically would prevent to have e.g. s32 to s16 conversion in IPC4 as well. It wasn't possible or working when I converted the module but now things have been fixed.

Reply:
Some explanation why IIR has format conversion:

In early days of SOF the preference for pipelines was 16 bits to save RAM in platforms like Baytrail. However in microphone paths if there was need to digitally boost the gain the quality was bad in topologies where capture DAI was 16 bit and we applied with volume or IIR about 20 dB gain. In practice a 16 bit word got left shifted by some bit positions that effectively made signal like 12 bits. We could achieve a lot better quality by capturing codec and DAI with 24 or 32bits and applying the gain in IIR for the larger word length. Then all 16 bits in the pipelines after DAI and IIR had signal. The IIR was chosen for format conversion because it also canceled the sometimes large DC component (and some lowest non-audible frequencies) in signal. It gave the headroom for signal for amplification.

If IPC4 systems ever need the memory save small 16 bit capture paths the format conversion could be brought back.

Reply:
thanks, will add